lib/Service/YAPServer.hpp

```diff
@@ -0,0 +1,594 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*******************************************************************************
+    DESCRIPTION
+*******************************************************************************/
+/**
+ * @brief  Yet Another Protocol (YAP) Server
+ * @author Gabryel Reyes <gabryelrdiaz@gmail.com>
+ *
+ * @addtogroup Service
+ *
+ * @{
+ */
+
+#ifndef YAP_SERVER_H
+#define YAP_SERVER_H
+
+/******************************************************************************
+ * Compile Switches
+ *****************************************************************************/
+
+/******************************************************************************
+ * Includes
+ *****************************************************************************/
+
+#include <YAPCommon.hpp>
+#include <Stream.h>
+#include <string.h>
+#include <Util.h>
+
+/******************************************************************************
+ * Macros
+ *****************************************************************************/
+
+/******************************************************************************
+ * Types and Classes
+ *****************************************************************************/
+
+/**
+ * Class for the YAP Server.
+ * @tparam tMaxChannels Maximum number of channels
+ */
+template<uint8_t tMaxChannels>
+class YAPServer
+{
+public:
+    /**
+     * Construct the YAP Server.
+     */
+    YAPServer(Stream& stream) :
+        m_isSynced(false),
+        m_lastSyncCommand(0U),
+        m_lastSyncResponse(0U),
+        m_pendingSuscribeChannel(),
+        m_stream(stream),
+        m_receiveFrame(),
+        m_receivedBytes(0U),
+        m_rxAttempts(0U)
+    {
+    }
+
+    /**
+     * Destroy the YAP Server.
+     */
+    ~YAPServer()
+    {
+    }
+
+    /**
+     * Manage the Server functions.
+     * Call this function cyclic.
+     * @param[in] currentTimestamp Time in milliseconds.
+     */
+    void process(const uint32_t currentTimestamp)
+    {
+        /* Periodic Heartbeat */
+        heartbeat(currentTimestamp);
+
+        /* Process RX data */
+        processRxData();
+    }
+
+    /**
+     * Send a frame with the selected bytes.
+     * @param[in] channelNumber Channel to send frame to.
+     * @param[in] payload Byte buffer to be sent.
+     * @param[in] payloadSize Amount of bytes to send.
+     */
+    void sendData(uint8_t channelNumber, const uint8_t* payload, uint8_t payloadSize)
+    {
+        if ((CONTROL_CHANNEL_NUMBER != channelNumber) && (nullptr != payload))

```

Comment:
What if channel number if max channels or higher?

Reply:
It is handled in the send() method, using getChannelDLC